src/test/java/se/sundsvall/billingdatacollector/service/DbServiceTest.java

```diff
@@ -0,0 +1,277 @@
+		verifyNoMoreInteractions(mockFalloutRepository);
+	}
+
+	@Test
+	void testSaveFailedBillingRecord_shouldNotSave_whenAlreadyExists() {
+		//Arrange
+		var wrapper = createBillingRecordWrapper();
+		when(mockFalloutRepository.existsByFamilyIdAndFlowInstanceIdAndBillingRecordWrapperIsNotNull(FAMILY_ID, FLOW_INSTANCE_ID)).thenReturn(true);
+
+		//Act
+		dbService.saveFailedBillingRecord(wrapper, "an error message");
+
+		//Assert
+		verify(mockFalloutRepository).existsByFamilyIdAndFlowInstanceIdAndBillingRecordWrapperIsNotNull(FAMILY_ID, FLOW_INSTANCE_ID);
+		verify(mockFalloutRepository, never()).saveAndFlush(Mockito.any(FalloutEntity.class));
+		verifyNoMoreInteractions(mockFalloutRepository);
+	}
+
+	@Test
+	void testSaveFailedFlowInstance() {
+		//Arrange
+		byte[] bytes = new byte[0];
+		when(mockFalloutRepository.existsByFamilyIdAndFlowInstanceIdAndOpenEInstanceIsNotNull(FAMILY_ID, FLOW_INSTANCE_ID)).thenReturn(false);
+		when(mockFalloutRepository.saveAndFlush(Mockito.any(FalloutEntity.class))).thenReturn(FalloutEntity.builder().build());
+
+		//Act
+		dbService.saveFailedFlowInstance(bytes, FLOW_INSTANCE_ID, FAMILY_ID, "an error message");
+
+		//Assert
+		verify(mockFalloutRepository).existsByFamilyIdAndFlowInstanceIdAndOpenEInstanceIsNotNull(FAMILY_ID, FLOW_INSTANCE_ID);
+		verify(mockFalloutRepository).saveAndFlush(argThat(entity ->
+			entity.getFamilyId().equals(FAMILY_ID) &&
+			entity.getFlowInstanceId().equals(FLOW_INSTANCE_ID) &&
+			entity.getOpenEInstance().equals(new String(bytes, StandardCharsets.ISO_8859_1))));
+		verifyNoMoreInteractions(mockFalloutRepository);
+	}
+
+	@Test
+	void testSaveFailedOpenEInstance_shouldNotSave_whenAlreadyExists() {
+		//Arrange
+		byte[] bytes = new byte[0];
+		when(mockFalloutRepository.existsByFamilyIdAndFlowInstanceIdAndOpenEInstanceIsNotNull(FAMILY_ID, FLOW_INSTANCE_ID)).thenReturn(true);
+
+		//Act
+		dbService.saveFailedFlowInstance(bytes, FLOW_INSTANCE_ID, FAMILY_ID, "an error message");
+
+		//Assert
+		verify(mockFalloutRepository).existsByFamilyIdAndFlowInstanceIdAndOpenEInstanceIsNotNull(FAMILY_ID, FLOW_INSTANCE_ID);
+		verify(mockFalloutRepository, never()).saveAndFlush(Mockito.any(FalloutEntity.class));
+		verifyNoMoreInteractions(mockFalloutRepository);
+	}
+
+	@Test
+	void testSaveToHistory() {
+		//Arrange
+		var wrapper = createBillingRecordWrapper();
+		var headers = createHeaders();
+
+		when(mockResponseEntity.getHeaders()).thenReturn(headers);
+
+		//Act
+		dbService.saveToHistory(wrapper, mockResponseEntity);
+
+		//Assert
+		verify(mockResponseEntity).getHeaders();
+		verify(mockHistoryRepository).saveAndFlush(argThat(entity ->
+			entity.getFlowInstanceId().equals(FLOW_INSTANCE_ID) &&
+			entity.getFamilyId().equals(FAMILY_ID) &&
+			entity.getLocation().equals("/location/uuid")));
+		verifyNoMoreInteractions(mockFalloutRepository);
+	}
+
+	@Test
+	void testHasAlreadyBeenProcessed_shouldReturnTrue_whenEitherIsTrue() {
```

Comment:
Consider making `testHasAlreadyBeenProcessed_shouldReturnTrue_whenEitherIsTrue` to a parameterized test (along with `testHasAlreadyBeenProcessed_shouldReturnFalse_whenNeitherIsTrue`) to verify all possible paths. Right now only path when both a true or if existsByFamilyIdAndFlowInstanceId is true are tested. Path where only existsByFamilyIdAndFlowInstanceId is true is never tested.

Reply:
Done!